### PR TITLE
Build using Stack / Cabal & add CI workflow

### PR DESCRIPTION
**Pull request by d3adb5**
These commits move the project's source code into a `src` directory to better
separate the Haskell files from unrelated ones (e.g. build files, Dockerfile).

Additionally, the files `stack.yaml` and `oshit.cabal` were added to allow
building and dependency management through Stack and Cabal respectively.

On top of this, for the purposes of continuous integration, a GitHub Actions
workflow was added to build the project using Stack. It takes longer than what
might be ideal, since GHC, Stack and the project's dependencies have to be
installed and, in the latter case, compiled, but Actions is currently free for
public repositories.

There were plenty of warnings thrown when the `-Wall` flag was used, and an
effort was made to eliminate them completely.

There are a few improvements I would like to suggest, that I might implement
myself in later pull requests:

- There is little to no use of typeclasses when they can be incredibly useful.
  For instance, `Core.Reflog` defines the data type `Reflog` and the functions
  `reflogToString` and `parseReflog`, but there are no instances of the `Show`
  and `Read` typeclasses for the `Reflog` type.

- There are some unnecessary anonymous functions thrown into the mix for the
  sake of pattern matching. While pattern matching is incredibly useful, the
  record syntax makes it so it is unnecessary in these cases and just makes the
  code more difficult to read. Compare:

  ```haskell
  data ExampleADT = Example
    { fieldOne :: String
    , fieldTwo :: Int
    }

  -- good, immediate use of pattern matching!
  matchPattern :: ExampleADT -> IO ()
  matchPattern (Example one _) = putStrLn one

  -- using a lambda function with pattern matching despite the records
  middleOfDoBlock :: [...] -> ExampleADT -> IO ()
  middleOfDoBlock ... = do
    -- ...
    let one = (\(Example one two) -> one) exampleADTinstance
    -- versus
    let one = fieldOne exampleADTinstance
    -- or using the above where the value is needed
  ```

  I caught a few instances of this and other unnecessary pattern matching with
  heavy name shadowing that was eliminated in c1e122c. I simplified some
  snippets, but left most of them as-is. The record syntax automatically makes
  it so functions `fieldOne :: ExampleADT -> String` and `fieldTwo :: ExampleADT
  -> Int` are created.

- I got the impression sometimes lists of `String` were overused, but I didn't
  familiarize myself with the codebase enough to be able to say this is as
  prevalent as the matter of pattern matching.

I like what you're doing with `oshit`, it's truly interesting! Wish I had more
time and Git know-how to be able to contribute to it beyond minor improvements.
One thing I would like to suggest, that I myself am not very familiar with, is
writing unit tests to be executed through Stack.

Also, given the maturity of Stack and Cabal, I feel like we can ditch GNU Make
as a build option.

### CODE REVIEW

**Comment by d3adb5** [comment]
Let me know what you think when you have the time to review these changes! I feel like the GitHub Actions workflow isn't being executed because I'm a new contributor and not a collaborator added to the project. If you'll have your workflow revolving around pull requests or just want to see whether `master`'s latest commit still builds, it's advisable to activate it.

There are no unit tests, and I didn't really take the time to test the minor changes with an actual Git repository, so it could be I introduced a few bugs! My changes were purely for eliminating warnings, and I'm going by the --- often wrong --- proverb of "if it compiles, it works."

**Comment by d3adb5** [comment]
Rebased and force-push'd. Had to fix a couple conflicts, and sorta unname some of your variables to eliminate unused-local-binding warnings. Sorry... :slightly_frowning_face:

**Review by lucasoshiro — APPROVED** [review]

**Comment by lucasoshiro** [comment]
> These commits move the project's source code into a `src` directory to better separate the Haskell files from unrelated ones (e.g. build files, Dockerfile).
> 
> Additionally, the files `stack.yaml` and `oshit.cabal` were added to allow building and dependency management through Stack and Cabal respectively.

Thanks! This is amazing!

Stack is working in my Manjaro / i5 2nd generation, however, it is not working in my Mac OS / Apple M1. It seems that this architecture is not supported yet... Cabal is working nicely!

File: src/Commands/Plumbing/Commit/CommitFromTree.hs
Comment (d3adb5) [reply src/Commands/Plumbing/Commit/CommitFromTree.hs:14]: I'd be hard pressed to say they are good practice anywhere, to be quite frank. While nigh ubiquitous (pick any random package from Hackage and upon inspecting its source code you're likely to find functions that have arguments with single letter names), I'd much rather have a descriptive name, even if a little long.

In this particular case `parents`, `author` and `email` sadly shadow the bindings created by using the record syntax, and thus generate warnings. For clarity when reading the code, it might be best not to name-shadow, so maybe take a page out of the Lens book and update the type definition somewhat like this?

```haskell
data Commit = Commit { _author :: String
                     , _email :: String
                     -- ...
                     }
```

What do you think?

**Comment by lucasoshiro** [comment]
> There are a few improvements I would like to suggest, that I might implement myself in later pull requests:
> 
> * There is little to no use of typeclasses when they can be incredibly useful.
>   For instance, `Core.Reflog` defines the data type `Reflog` and the functions
>   `reflogToString` and `parseReflog`, but there are no instances of the `Show`
>   and `Read` typeclasses for the `Reflog` type.

Ok, in my TODO list (#1)

> The record syntax automatically makes
>   it so functions `fieldOne :: ExampleADT -> String` and `fieldTwo :: ExampleADT -> Int` are created.

I didn't know about that, cool. This is also going to my TODO list.

> * I got the impression sometimes lists of `String` were overused, but I didn't
>   familiarize myself with the codebase enough to be able to say this is as
>   prevalent as the matter of pattern matching.

`git` stores most of things as strings, for example:
- branches, tags and other references are just files with a SHA-1 string;
- everything inside a commit object (author, committer, parents and snapshot hashes and message) is string;
- symbolic references are files with a SHA-1 string or a branch name;

Some exceptions are the index file and the `tree` objects, that contain some binary data. I also think that I'm using too much `String`s and `ByteString`s, but I think the most I can in order to solve it add some typealiases...

**Comment by d3adb5** [comment]
I see you approved the changes, but before a merge, is it possible to enable the GitHub Actions workflow? That way we can keep track of CI on GitHub should any further pull requests be opened. Is this repository being mirrored on GitLab? Do you prefer GitLab CI over GitHub Actions or...?

Just wondering, I don't know what your stance on Actions or host-provided CI/CD is.

**Comment by d3adb5** [comment]
> Some exceptions are the index file and the tree objects, that contain some binary data. I also think that I'm using too much `String`s and `ByteString`s, but I think the most I can in order to solve it add some typealiases...

Not a big fan of type aliases, to be quite frank. If it's inevitable to tie objects together this way, so be it. I was hoping each object in Git might be correlated 1:1 with a data type here. If that's already being done, guess I'm just inattentive.

> > There is little to no use of typeclasses when they can be incredibly useful.
For instance, Core.Reflog defines the data type Reflog and the functions
reflogToString and parseReflog, but there are no instances of the Show
and Read typeclasses for the Reflog type.
>
> Ok, in my TODO list (https://github.com/lucasoshiro/oshit/issues/1)

I started the work on adding tests and typeclass instances in #3. We can work on it together, if you're interested! I've been dusting off my Haskell skills and picking up writing unit tests with Hspec and QuickCheck. What do you think?

**Comment by lucasoshiro** [comment]
> I like what you're doing with `oshit`, it's truly interesting! Wish I had more time and Git know-how to be able to contribute to it beyond minor improvements.

Thanks for your time and work here @d3adb5 ! This is just a side project that I'm coding as I learn new things about `git`. Some of the decisions so far are wrong (such as defining `author` and `committer` as the same) as I am too lazy to them correctly... And some of them are because I think it makes more sense than the original ones, e. g.: `oshit create-blob` makes more sense to me than `git hash-object --stdin -w`

**Comment by lucasoshiro** [comment]
>  One thing I would like to suggest, that I myself am not very familiar with, is writing unit tests to be executed through Stack.

I was writing some tests locally using shellscript, just like `git` does (e. g.: https://github.com/git/git/blob/master/t/t2405-worktree-submodule.sh). I gave up 😁. But I'll check this out. This project really needs some unit tests.

> Also, given the maturity of Stack and Cabal, I feel like we can ditch GNU Make as a build option.

Sure!

**Comment by lucasoshiro** [comment]
> I see you approved the changes, but before a merge, is it possible to enable the GitHub Actions workflow? 

I think I can't do that as this branch is not mine...

However, in this #4 I pushed you branch to this repo, the workflow worked, and as it was the same commit, this PR also passed in that check, as you can see.

**Comment by lucasoshiro** [comment]
> I was hoping each object in Git might be correlated 1:1 with a data type here. If that's already being done, guess I'm just inattentive.

It is, we have the types `Tree`, `Commit` and `Blob`, declared [here](https://github.com/git/git/blob/c2162907e9aa884bdb70208389cb99b181620d51/merge-ort.c#L1912-L1915). But most of the time, we don't need to know it's content, their hashes are enough. For example, take a look at the tree-way merge, it handles the merge of files only looking the hash of the blobs that contain its contents:

https://github.com/lucasoshiro/oshit/blob/d2280c7e7a65a7c2b443052fc4a3954f00de7eda/Core/Merge.hs#L15-L20

(as a curiosity, `git` also does that: https://github.com/git/git/blob/c2162907e9aa884bdb70208389cb99b181620d51/merge-ort.c#L1912-L1915)

When we need to read objects, I'm using `ObjectIO` (declared [here](https://github.com/lucasoshiro/oshit/blob/d2280c7e7a65a7c2b443052fc4a3954f00de7eda/Core/Object.hs#L39-L43)). This takes advantage of how IO works in Haskell, only reading the object content if needed.

**Comment by lucasoshiro** [comment]
> I started the work on adding tests and typeclass instances in #3. We can work on it together, if you're interested! I've been dusting off my Haskell skills and picking up writing unit tests with Hspec and QuickCheck. What do you think?

Nice, thanks again!

**Comment by d3adb5** [comment]
> > I see you approved the changes, but before a merge, is it possible to enable the GitHub Actions workflow?
> 
> I think I can't do that as this branch is not mine...
> 
> However, in this #4 I pushed you branch to this repo, the workflow worked, and as it was the same commit, this PR also passed in that check, as you can see.

That's odd. Oh well. All's well that ends well. I thought it'd have something to do with the Actions configuration. :man_shrugging:

**Pull request merged**
